plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java

```diff
@@ -32,16 +34,63 @@
 import org.eclipse.jface.wizard.IWizard;
 import org.eclipse.m2e.core.MavenPlugin;
 import org.eclipse.m2e.core.internal.IMavenConstants;
+import org.eclipse.m2e.core.internal.lifecyclemapping.discovery.LifecycleMappingDiscoveryRequest;
 import org.eclipse.m2e.core.internal.project.ProjectConfigurationManager;
 import org.eclipse.m2e.core.project.IProjectConfigurationManager;
 import org.eclipse.m2e.core.project.LocalProjectScanner;
 import org.eclipse.m2e.core.project.MavenProjectInfo;
 import org.eclipse.m2e.core.project.MavenUpdateRequest;
 import org.eclipse.m2e.core.project.ResolverConfiguration;
+import org.eclipse.m2e.core.ui.internal.wizards.LifecycleMappingDiscoveryHelper;
+import org.eclipse.m2e.core.ui.internal.wizards.MappingDiscoveryJob;
+import org.eclipse.ui.internal.wizards.datatransfer.EasymportJob;
 import org.eclipse.ui.wizards.datatransfer.ProjectConfigurator;
 
 public class MavenProjectConfigurator implements ProjectConfigurator {
 	
+	private static class CumulativeMappingDiscoveryJob extends MappingDiscoveryJob {
+		private static CumulativeMappingDiscoveryJob INSTANCE;
+		private Set<IProject> toProcess;
+		
+		public synchronized static CumulativeMappingDiscoveryJob getInstance() {
+			if (INSTANCE == null) {
+				INSTANCE = new CumulativeMappingDiscoveryJob();
+			}
+			return INSTANCE;
+		}
+		
+		private CumulativeMappingDiscoveryJob() {
+			super(null);
+			this.toProcess = new HashSet<IProject>();
+		}
+		
+		public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
+			try {
+				getJobManager().join(EasymportJob.class, monitor);
+			} catch (InterruptedException ex) {
+				throw new CoreException(new Status(IStatus.WARNING, Activator.getDefault().getBundle().getSymbolicName(), ex.getMessage(), ex));
+			}
+			//Detect and resolve Lifecycle Mapping issues
+			LifecycleMappingDiscoveryRequest discoveryRequest = LifecycleMappingDiscoveryHelper.createLifecycleMappingDiscoveryRequest(toProcess, monitor);
+			if(discoveryRequest.isMappingComplete()) {
+				return Status.OK_STATUS;
+			}
+			//Some errors were detected
+			discoverProposals(discoveryRequest, monitor);
+			openProposalWizard(toProcess, discoveryRequest);
+			this.toProcess.clear();
+			return Status.OK_STATUS;
+		}
+			
+		public void addProjects(Collection<IProject> projects) {
+			if (projects != null) {
```

Comment:
I think this method should check the job status before adding new projects. If the instance is already running, gets past openProposalWizard(), projects will be added to the set that'll be cleared right after.
So basically you should throw an IllegalStateException if the job is running while you're adding projects